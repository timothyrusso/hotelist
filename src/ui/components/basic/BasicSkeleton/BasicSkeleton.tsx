import { Animated, type StyleProp, type ViewStyle } from "react-native";
import { useBasicSkeletonLogic } from "./BasicSkeleton.logic";

interface BasicSkeletonProps {
	children?: React.ReactNode;
	style?: StyleProp<ViewStyle>;
}

export const BasicSkeleton: React.FC<BasicSkeletonProps> = ({
	children,
	style,
}) => {
	const { styles } = useBasicSkeletonLogic();

	return (
		<Animated.View style={[styles.container, style]}>{children}</Animated.View>
	);
};
