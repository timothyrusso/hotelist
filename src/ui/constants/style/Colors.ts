export const Colors = {
	alpha: {
		default: "#C2DFFF",
		hover: "#A9C8F0",
		disabled: "#E3ECF8",
	},
	beta: {
		default: "#FFE2A1",
		hover: "#FFD385",
		disabled: "#F6E8C7",
	},
	gamma: {
		default: "#B7B7FF",
		hover: "#9E9EFF",
		disabled: "#D7D7F5",
	},
	delta: {
		default: "#D8E9AE",
		hover: "#C6DB94",
		disabled: "#E7F1D2",
	},
	black: {
		default: "#1A1A1A",
		hover: "#1A1A1A",
		disabled: "#B8B5B1",
	},
	white: {
		default: "#FFFFFF",
		hover: "#F0F0F0",
		disabled: "#F0F0F0",
	},
} as const;
