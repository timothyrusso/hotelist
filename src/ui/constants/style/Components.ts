export const Components = {
	hotelCard: {
		image: {
			width: 100,
		},
	},
	hotelDetailsPage: {
		image: {
			height: 300,
		},
	},
};
