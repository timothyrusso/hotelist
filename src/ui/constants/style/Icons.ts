export const Icons = {
	star: "star",
	hotel: "hotel",
	location: "location-sharp",
	euro: "euro",
	search: "search",
	options: "options",
	sad: "sad",
	close: "close-a",
	chevronRight: "chevron-right",
	filter: "filter",
} as const;
