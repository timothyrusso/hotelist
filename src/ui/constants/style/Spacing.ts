export const Spacing = {
	Zero: 0,
	HalfMinimal: 1,
	ThreeQuarterMinimal: 1.5,
	Minimal: 2,
	MinimalDouble: 4,
	Single: 6,
	SingleAndHalf: 8,
	Double: 12,
	Triple: 16,
	TripleAndHalf: 20,
	Fourfold: 24,
	Quintuple: 32,
	Sextuple: 48,
	separator40: 40,
	separator80: 80,
	separator120: 120,
	separator160: 160,
};
