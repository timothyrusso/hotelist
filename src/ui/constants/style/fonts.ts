export const fonts = {
	interRegular: "inter-regular",
	interMedium: "inter-medium",
	interBold: "inter-bold",
} as const;

export const fontsConfig = {
	"inter-regular": require("../../assets/fonts/Inter-Regular.ttf"),
	"inter-medium": require("../../assets/fonts/Inter-Medium.ttf"),
	"inter-bold": require("../../assets/fonts/Inter-Bold.ttf"),
};
