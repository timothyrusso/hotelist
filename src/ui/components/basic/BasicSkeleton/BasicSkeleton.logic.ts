import { useEffect, useRef } from "react";
import { Animated } from "react-native";
import { styles as skeletonStyles } from "./BasicSkeleton.style";

const PULSE_DURATION = 800; // Duration for each direction of the pulse (in milliseconds)

const PULSE_STARTING_OPACITY = 0.4; // Minimum opacity (dimmer state)
const PULSE_ENDING_OPACITY = 1.0; // Maximum opacity (brighter state)

export const useBasicSkeletonLogic = () => {
	// Create an animated value for opacity, starting at the minimum opacity
	// useRef ensures the animated value persists across re-renders
	const opacityAnim = useRef(
		new Animated.Value(PULSE_STARTING_OPACITY),
	).current;

	useEffect(() => {
		// Create a looping animation sequence that pulses the opacity
		const pulseAnimation = Animated.loop(
			Animated.sequence([
				// Step 1: Animate from starting opacity to ending opacity (fade in)
				Animated.timing(opacityAnim, {
					toValue: PULSE_ENDING_OPACITY,
					duration: PULSE_DURATION,
					useNativeDriver: true,
				}),
				// Step 2: Animate from ending opacity back to starting opacity (fade out)
				Animated.timing(opacityAnim, {
					toValue: PULSE_STARTING_OPACITY,
					duration: PULSE_DURATION,
					useNativeDriver: true,
				}),
			]),
		);

		// Start the animation loop
		pulseAnimation.start();

		// Cleanup function: stop the animation when component unmounts
		// This prevents memory leaks and ensures the animation stops properly
		return () => {
			pulseAnimation.stop();
		};
	}, [opacityAnim]); // Dependency array: re-run effect if opacityAnim changes

	const styles = skeletonStyles(opacityAnim);

	return {
		styles,
	};
};
