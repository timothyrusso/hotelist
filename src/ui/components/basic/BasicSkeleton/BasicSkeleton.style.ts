import { Colors } from "@/src/ui/constants/style/Colors";
import { type Animated, StyleSheet } from "react-native";

export const styles = (opacity: Animated.Value) =>
	StyleSheet.create({
		container: {
			backgroundColor: Colors.black.disabled,
			opacity: opacity,
		},
	});
