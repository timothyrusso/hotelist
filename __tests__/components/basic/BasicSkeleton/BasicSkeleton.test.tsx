import { BasicSkeleton } from "@/src/ui/components/basic/BasicSkeleton/BasicSkeleton";
import { Spacing } from "@/src/ui/constants/style/Spacing";
import { render } from "@testing-library/react-native";

describe("BasicSkeleton", () => {
	it("should render correctly with default props", () => {
		const { toJSON } = render(<BasicSkeleton />);

		// Verify the component renders correctly
		expect(toJSON()).toMatchSnapshot();
	});

	it("should render correctly with children", () => {
		const { toJSON } = render(
			<BasicSkeleton>
				<BasicSkeleton />
			</BasicSkeleton>,
		);

		expect(toJSON()).toMatchSnapshot();
	});

	it("should render correctly with custom style", () => {
		const { toJSON } = render(
			<BasicSkeleton
				style={{ width: Spacing.separator120, height: Spacing.separator80 }}
			/>,
		);

		expect(toJSON()).toMatchSnapshot();
	});

	it("should render correctly with children and custom style", () => {
		const { toJSON } = render(
			<BasicSkeleton style={{ padding: Spacing.Triple }}>
				<BasicSkeleton style={{ margin: Spacing.Double }} />
			</BasicSkeleton>,
		);

		expect(toJSON()).toMatchSnapshot();
	});
});
