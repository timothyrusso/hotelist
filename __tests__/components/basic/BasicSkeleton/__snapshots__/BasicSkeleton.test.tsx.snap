// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BasicSkeleton should render correctly with children 1`] = `
<View
  collapsable={false}
  style={
    {
      "backgroundColor": "#e8e8e8",
      "opacity": 0.4,
    }
  }
>
  <View
    collapsable={false}
    style={
      {
        "backgroundColor": "#e8e8e8",
        "opacity": 0.4,
      }
    }
  />
</View>
`;

exports[`BasicSkeleton should render correctly with children and custom style 1`] = `
<View
  collapsable={false}
  style={
    {
      "backgroundColor": "#e8e8e8",
      "opacity": 0.4,
      "padding": 16,
    }
  }
>
  <View
    collapsable={false}
    style={
      {
        "backgroundColor": "#e8e8e8",
        "margin": 12,
        "opacity": 0.4,
      }
    }
  />
</View>
`;

exports[`BasicSkeleton should render correctly with custom style 1`] = `
<View
  collapsable={false}
  style={
    {
      "backgroundColor": "#e8e8e8",
      "height": 80,
      "opacity": 0.4,
      "width": 120,
    }
  }
/>
`;

exports[`BasicSkeleton should render correctly with default props 1`] = `
<View
  collapsable={false}
  style={
    {
      "backgroundColor": "#e8e8e8",
      "opacity": 0.4,
    }
  }
/>
`;
